### ISSUE TITLE
Have pgr_version register the commit that generated the binaries

### ISSUE THREAD

**Issue by robe2** [pgRouting/pgrouting#342]
I don't know why I thought this was the case before.  I thought the hash was based on the git version because compiled, but it seems develop is stuck at: 5cdd106

```
 version |       tag       | build |  hash   | branch  | boost
---------+-----------------+-------+---------+---------+--------
 2.0.0   | pgrouting-2.0.0 | 170   | 5cdd106 | develop | 1.53.0
(1 row)
```
Which is what @woodbri  also read before the mega ksp commit.

From a testing perspective -- particularly because I build for windows users and winnie builds automatically, I can't always tell I am testing the recent revision and which version people are running when they come to me with issues (if they pulled from winnie).  I'm not sure how hard this is to do, but would be really nice.

For context on the PostGIS side of things the 

```
SELECT postgis_full_version();

POSTGIS="2.2.0dev r13510" GEOS="3.5.0dev-CAPI-1.9.0 r4054" PROJ="Rel. 4.8.0, 6 March 2012" GDAL="GDAL 1.11.1, released 2014/09/24" LIBXML="2.7.8" LIBJSON="0.12" RASTER
```

contains the revision number -- that way I can always trace back an install to the commit that generated it.

**Comment by dkastl** [comment]
I think there are 2 issues with the way we do this right now:

1. One has to copy the "pre-commit" hook manually after cloning the repository: https://github.com/pgRouting/pgrouting/blob/master/tools/pre-commit. I must admit, that I also forgot this already after I made a new clone.
2. The hash is not known until the commit is done, so you are always one commit behind or need to go a `git commit --amend` for the correct hash. There is no way to change this with "pre-commit" hooks, but maybe someone knows a better way to do this.

To summarize, the current way it works is error-prone.

**Comment by cvvergara** [comment]
So this last "mega ksp commit"  ;)  needs another hash number even if it is behind?

**Comment by boerngen-schmidt** [comment]
For me this sound more like a `post-checkout` hook. So basically you want to know what is the current hash of the checkout, which will build, right? So after a checkout a file can be written with the current hash. This value will stay the same on releases, since they are bound to a certain hash and will change for each development version.

**Comment by cvvergara** [comment]
We are one has behind, but we are getting different hash number now.
We will welcome a pull request that doesn't make us be a hash number behind.

**Comment by dkastl** [comment]
How do you want to know the commit ID before you did the commit?

The post-checkout hook looks interesting,

**Comment by woodbri** [comment]
There is NO way to get the commit hash until after it has been committed because the has is a function of the commit. And you can not edit the committed file after the commit because that makes a new version of the file that has not been committed. I would be happy to review a pull request if you think you have some way to do this.

**Comment by dkastl** [comment]
This is a good article about Git hooks: https://www.atlassian.com/git/tutorials/git-hooks

The important thing is, that you can't force a developer to "install" hooks. Each developer is responsible to have them in place. There are server-side hooks as well, but we can only use, what Github provides us.

So eventually there is a better way to automatically run `git --amend` after a commit to include the latest hash in the version file, but Git hooks have their limitations.

I'm curious how other projects manage version hashes or similar stuff.

**Comment by boerngen-schmidt** [comment]
Just an idea that came into my mind is the usage of the [keyword extension](https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes#Keyword-Expansion) $Id$. On SVN the $Id$ was updated with a string indicating the revision number of the current checkout. So having this keyword in the pgr_version implementation hardcoded and replaced on checkout by git would give you the hash of the current checkout (I haven't tried it though)

**Comment by dkastl** [comment]
@boerngen-schmidt if that would work, that would be great. I also found this keyword extension, but as I read it, it won't be the commit hash but an ID, and I'm not sure it's the same for any file. And also it will not tell us which commit it is, which is the main purpose of pgr_version() function.

**Comment by boerngen-schmidt** [comment]
@dkastl sadly it does not work the way I expected it to work. So the keyword extension is useless. But my brain dug out something new. The build system! :-) and google found a very similar question http://stackoverflow.com/a/1796675/4303718

When thinking about it, the build system is the best way to do this, because one wants to know on which checkout the build is based on, right?

**Comment by cvvergara** [comment]
I'll put this into milestone 2.2, at least for 2.1 we are getting a different hash even if its one hash behind

**Comment by cvvergara** [comment]
moving this further, we aare still on hash behind

**Comment by dkastl** [comment]
In general good, but shouldn'T we maybe add at least into one example a value for `alpha`?
Just an idea.

**Comment by cvvergara** [comment]
yes, and I what happens when the update is done using the git web page, the VERSION file doesn't get updated either, and I think when I did the mega ksp I didnt know about the hook, but I use it on a daily  basis now so, if we know we are a hash behind, we are being consistent.

**Comment by cvvergara** [comment]
I'll close it.
